string.c

```diff
@@ -343,9 +348,28 @@ const struct st_hash_type rb_fstring_hash_type = {
 
 #define BARE_STRING_P(str) (!FL_ANY_RAW(str, FL_EXIVAR) && RBASIC_CLASS(str) == rb_cString)
 
+static VALUE
+str_precompute_hash(VALUE str)
+{
+    RUBY_ASSERT(!FL_TEST_RAW(str, STR_PRECOMPUTED_HASH));
+    RUBY_ASSERT(STR_EMBED_P(str));
+
+#if RUBY_DEBUG
+    size_t used_bytes = (RSTRING_LEN(str) + TERM_LEN(str));
+    size_t free_bytes = str_embed_capa(str) - used_bytes;
+    RUBY_ASSERT(free_bytes >= sizeof(st_index_t));
+#endif
+
+    *(st_index_t *)(RSTRING_END(str) + TERM_LEN(str)) = rb_str_hash(str);
```

Comment:
I am afraid if this may cause a bus error on some platform. Is it safe?

Reply:
> this may cause a bus error on some platform

Because of alignment? If it's a concern, I think we could ensure it's always aligned, the overhead would be very slightly larger.

Reply:
Yes > because of alignment

It would be safer to use aligned address if possible.

Reply:
Thanks, I'll look into it.

Reply:
```suggestion
    typedef struct {char bytes[sizeof(st_index_t)];} unaligned_index;
    union {st_index_t i; unaligned_index b;} u = {.i = rb_str_hash(str)};
    *(unaligned_index *)(RSTRING_END(str) + TERM_LEN(str)) = u.b;
```

Reply:
Thank @nobu, interesting trick. I suppose we need it on read as well?